SpamPkg/Core/Runtime/SpamResponderReport.c

```diff
@@ -0,0 +1,889 @@
+
+  // At this point we dealt with the relocation, some data are still off.
+  // Next we unload the image in the copy.
+  Status = PeCoffLoaderRevertLoadImage (&ImageContext, NewBuffer, &NewBufferSize);
+  if (EFI_ERROR (Status)) {
+    goto Exit;
+  }
+
+  DEBUG ((DEBUG_INFO, "%a Reverted image at %p of size %x\n", __func__, NewBuffer, NewBufferSize));
+  ASSERT (MmSupvEfiFileSize == NewBufferSize);
+  ASSERT (CompareMem (NewBuffer, (VOID*)(UINTN)MmSupvEfiFileBase, MmSupvEfiFileSize) == 0);
+
+  Status = MeasurePeImageAndExtend (
+              SPAM_PCR_INDEX,
+              (EFI_PHYSICAL_ADDRESS)(UINTN)NewBuffer,
+              (UINTN)NewBufferSize,
+              DigestList
+              );
+  if (!EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_INFO, "%a Measured image at %p of size %x successfully.\n", __func__, NewBuffer, NewBufferSize, Status));
+  } else {
+    DEBUG ((DEBUG_ERROR, "%a Failed to measure image at %p of size %x - %r\n", __func__, NewBuffer, NewBufferSize, Status));
+  }
+
+Exit:
+  if (InternalCopy != NULL) {
+    FreePages (InternalCopy, EFI_SIZE_TO_PAGES (ImageSize + EFI_PAGE_SIZE - 1));
+  }
+  return Status;
+}
+
+EFI_STATUS
+EFIAPI
+SpamResponderReport (
+  IN SPAM_RESPONDER_DATA *SpamResponderData
+  )
+{
+  EFI_STATUS                      Status;
+  UINT64                          MmBase;
+  UINT64                          MmRamBase;
+  UINT64                          MmrrMask;
+  UINT32                          MaxExtendedFunction;
+  CPUID_VIR_PHY_ADDRESS_SIZE_EAX  VirPhyAddressSize;
+  UINT64                          Length;
+  UINT64                          MtrrValidBitsMask;
+  UINT64                          MtrrValidAddressMask;
+  UINT16                          *FixStructPtr;
+  UINT8                           *Fixup8Ptr;
+  UINT32                          *Fixup32Ptr;
+  UINT64                          *Fixup64Ptr;
+  BOOLEAN                         IsInside;
+  UINTN                           Index;
+  PER_CORE_MMI_ENTRY_STRUCT_HDR   *MmiEntryStructHdr;
+  UINT32                          MmiEntryStructHdrSize;
+  UINT64                          MmSupervisorBase;
+  UINT64                          FirmwarePolicyBase;
+  UINT64                          SupvPageTableBase;
+  TPML_DIGEST_VALUES              DigestList[HASH_COUNT];
+
+  KEY_SYMBOL                      *FirmwarePolicySymbol = NULL;
+  KEY_SYMBOL                      *PageTableSymbol = NULL;
+  KEY_SYMBOL                      *MmiRendezvousSymbol = NULL;
+
+  // TODO: Step 0: Disable MMI
+
+  // Step 1: Basic check on the validity of SpamResponderData
+  // TODO: How do we know if the input stack is safe to access?
+  if (SpamResponderData == NULL) {
+    Status = EFI_INVALID_PARAMETER;
+    goto Exit;
+  }
+
+  if (SpamResponderData->Signature != SPAM_RESPONDER_STRUCT_SIGNATURE) {
+    Status = EFI_UNSUPPORTED;
+    goto Exit;
+  }
+
+  if (SpamResponderData->VersionMajor > SPAM_REPSONDER_STRUCT_MAJOR_VER) {

```

Comment:
So this check seems to imply backwards compatibility.  Just want to confirm that's what you want.

Reply:
This basically implies that if there is a major version change, this version of spam core is unable to handle the incoming data. So we should not even try to proceed.

Reply:
The next case implies that if the change is minor update, we may still have a chance to parse the data, although there might be some data fields unable to be processed.